### PR TITLE
When updating ThreadLocalRandom used the old sun.misc.Unsafe rather than jdk.internal.misc.Unsafe

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: src/java/org/apache/cassandra/utils/reflection/ReflectionUtils.java
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/utils/reflection/ReflectionUtils.java]: Yup, I suggested this also in my branch for jdk 17 https://github.com/apache/cassandra/commit/4daf0b149bc1524dfb2212ab958f4d45feb79577#diff-929245717cd8c37ec78fb97bddc5b6839b2430c3e32f31f30e4aad5fcd580c71R376-R404 I will put a note to get rid of this in my WIP branch and use this utility class on next rebase :-) Thanks

File: build.xml
Comment (ekaterinadimitrova2) [inline build.xml]: why do we add now 17 to build xml?
? It can't even compile with it as Benjamin is not ready with the UDF refactoring.... Also, I already had the changes to make It 11 and 17, double work

File: build.xml
Comment (ekaterinadimitrova2) [inline build.xml]: Are those justified? I was looking for ways not to add openings.

File: test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java
Comment (ekaterinadimitrova2) [inline test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java:152]: This will be only for the simulator ?

File: build.xml
Comment (belliottsmith) [reply build.xml]: Perhaps we can add these on the simulator launch path only, to avoid taking on new add-opens just to unblock it.

File: build.xml
Comment (belliottsmith) [reply build.xml]: We don't need to do as complete a job as you, as we want to get JDK17 (actually JDK19) supported for the simulator only.

File: test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java
Comment (belliottsmith) [reply test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java:152]: Yep, this is in the simulator tree so I don't believe it's even available at compilation or runtime to other parts of the codebase.

Comment (dcapwell) [reply test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java:152]: yes, this is part of the simulator; we rewrite classes so we can control execution order... but on JDK 11 how private fields/methods are shard with subclasses changed, so rather than trying to make things complex, the rewrite logic makes everything public

File: src/java/org/apache/cassandra/utils/reflection/ReflectionUtils.java
Comment (dcapwell) [reply src/java/org/apache/cassandra/utils/reflection/ReflectionUtils.java]: reverted for now as this is for JDK 17

File: test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java
Comment (ekaterinadimitrova2) [reply test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java:152]: Yup, that's what I thought, thanks. For testing and being only used in the simulator sounds reasonable to me

File: build.xml
Comment (ekaterinadimitrova2) [inline build.xml:172]: Do we need to add some comment or something for this condition for completeness or I am tired and this will be obvious and I should leave you alone? :D

Comment (dcapwell) [reply build.xml:172]: ill add comment

**Review by belliottsmith — APPROVED** [review]

File: .circleci/config-2_1.yml
Comment (ekaterinadimitrova2) [inline .circleci/config-2_1.yml:1167]: Ok, so as we talked in Slack probably we can skip j8 build with j11 runtime and directly go for j11 build with j17 when we do the switch on trunk. I leave the decision to you whether you will add now j8/j11 or not. But don't we need this also added in the pre-commit workflow, same as all other test suites?

File: build.xml
Comment (ekaterinadimitrova2) [inline build.xml:157]: Thanks for adding this, it will be a good reminder in time as people might miss it!

File: build.xml
Comment (ekaterinadimitrova2) [inline build.xml]: Why this change was needed?Isn't this the same issue as earlier? I would prefer if we skip changes that are not explicitly needed. Did I miss anything here?

File: build.xml
Comment (ekaterinadimitrova2) [inline build.xml]: Why this change of name?

File: test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java
Comment (ekaterinadimitrova2) [inline test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java]: Not only methods? ACC_SYNTHETIC applies also to fields?

**Review by ekaterinadimitrova2 — COMMENTED** [review]
Overall +1, just some nits. Main questions about CircleCI - adding the job also to the pre-commit even if it still requires approval step. All jobs are in both workflows - the separate jobs one and the precommit one.
And a change in build.xml that I guess it was just forgotten to revert maybe when reverting the other jdk17 stuff.

So I think this needs squash, rebase and final run. We need to test also that if jobs are added or changed, the generate.sh will pick them up from diff as others and they will be pushed in a loop in CI to check for flakies :-) As there were many circle ci config changes (adding all missing jobs) in preparation for 4.1, the circleci files might be annoying. Let me know if you want me to land a hand there

File: .circleci/config-2_1.yml
Comment (dcapwell) [reply .circleci/config-2_1.yml:1167]: > But don't we need this also added in the pre-commit workflow, same as all other test suites?

:sigh: this build is getting too complicated... ill add

> we can skip j8 build with j11 runtime and directly go for j11 build with j17 

for now, since we support 8/11 ill keep the build doing 8/11; when we drop 8 and add 17 we can switch

Comment (dcapwell) [reply .circleci/config-2_1.yml:1167]: I wrote a script to confirm the change and found that if I manually applied on my own I would have messed up!

Here is the full diff of the workflows

```
Diff for ('java8_separate_tests', 'java11_separate_tests') is
left:
- start_java_dtest_jars_build
- java_dtest_jars_build
- start_jvm_upgrade_dtests
- java_jvm_upgrade_dtests
- start_jvm_upgrade_dtests_repeat
- java_jvm_upgrade_dtests_repeat
- start_upgrade_dtests
- java_upgrade_dtests
- start_java_cqlsh_tests_offheap
- start_java_upgrade_dtests_repeat
- java_upgrade_dtests_repeat
mismatch:
  java_cqlsh_dtests_py38_offheap:
    left:
    - start_java_cqlsh_tests_offheap
    - java_build
    right:
    - start_java_cqlsh-dtests-offheap
    - java_build
  java_cqlsh_dtests_py3_offheap:
    left:
    - start_java_cqlsh_tests_offheap
    - java_build
    right:
    - start_java_cqlsh-dtests-offheap
    - java_build
right:
- start_java_cqlsh-dtests-offheap

Diff for ('java8_pre-commit_tests', 'java11_pre-commit_tests') is
left:
- start_jvm_upgrade_dtests
- java_dtest_jars_build
- java_jvm_upgrade_dtests
- java_jvm_upgrade_dtests_repeat
- start_upgrade_dtests
- java_upgrade_dtests
- java_upgrade_dtests_repeat
- start_java_cqlsh_dtests_offheap
mismatch:
  java_dtests_offheap_repeat:
    left:
    - start_java_dtests_offheap
    - java_build
    right:
    - start_java_dtests_offheap_repeat
    - java_build
right:
- start_java_dtests_offheap_repeat

Diff for ('java8_separate_tests', 'java8_pre-commit_tests') is
left:
- j8_repeated_ant_test
- j11_repeated_ant_test

Diff for ('java11_separate_tests', 'java11_pre-commit_tests') is
left:
- j11_repeated_ant_test
```

Excluding differences not caused by this patch

```
$ circleci-verify.py --exclude upgrade dtest_jars cqlsh repeated_ant dtests_offheap
$
```

Here is the diff without the latest patch

```
$ circleci-verify.py --exclude upgrade dtest_jars cqlsh repeated_ant dtests_offheap
Diff for ('java8_pre-commit_tests', 'java11_pre-commit_tests') is
left:
- java_simulator_dtests
- java_simulator_dtests_repeat

Diff for ('java11_separate_tests', 'java11_pre-commit_tests') is
left:
- j11_simulator_dtests
- j11_simulator_dtests_repeat

```

File: test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java
Comment (dcapwell) [reply test/simulator/asm/org/apache/cassandra/simulator/asm/ClassTransformer.java]: updated comment

**Comment by dcapwell** [comment]
> And a change in build.xml that I guess it was just forgotten to revert maybe when reverting the other jdk17 stuff.

nope, build.xml is expected.  `arch_x86` is needed to build on j8 on M1 Macs, and the rest are cleanups (optional); stuff like adding better error message when you run on a unsupported JDK

**Comment by dcapwell** [comment]
> So I think this needs squash, rebase and final run

pushed latest changes

> We need to test also that if jobs are added or changed, the generate.sh will pick them up from diff as others and they will be pushed in a loop in CI to check for flakies 

now that the repeat jobs got deleted by default I don't know how to run them so I never run them anymore... how do you run it now?

File: build.xml
Comment (dcapwell) [reply build.xml]: its not needed, its duplicate work; why reinvent the wheel when we already have one?  I can revert if you desire; its only cleanup while I was here.

File: build.xml
Comment (dcapwell) [reply build.xml]: optional cleanup, can revert if you want; this was because JDK 17 but left as I felt it was fine

File: .circleci/config-2_1.yml
Comment (ekaterinadimitrova2) [reply .circleci/config-2_1.yml:1167]: The issue you caught is naming which is known (tests vs dtests and _ vs -) and planned to be fixed across the board at once in a separate ticket. The only thing I noticed is that one of the repeated tests jobs will always be started in pre-commit if tests to run in a loop are added for that suite which can be made to match the rest, it is inconsistency but not some huge bug considering how many legit bugs(and I do not mean just missing tests) I found and how many tests I added.

`:sigh: this build is getting too complicated... ill add`
Well, we finally have all possible tests in CI and thanks to that we should be able to release. Also, now when it is finally clear that no one anymore uses the HIGHRES as is, we can switch to two configs - paid and free. I am looking forward to those changes. Let's see when people will have the time.

File: build.xml
Comment (ekaterinadimitrova2) [reply build.xml]: Please revert, it is more clear when the version is mentioned, thanks

File: build.xml
Comment (ekaterinadimitrova2) [reply build.xml]: I don't follow...sorry... what is duplicate work?

File: .circleci/config-2_1.yml
Comment (dcapwell) [reply .circleci/config-2_1.yml:1167]: > paid and free. I am looking forward to those changes. Let's see when people will have the time.

doesn't address the comment, but +1 to unify MID and HIGHER to a PAID

File: build.xml
Comment (dcapwell) [reply build.xml]: ill revert

File: .circleci/config-2_1.yml
Comment (ekaterinadimitrova2) [reply .circleci/config-2_1.yml:1167]: [CASSANDRA-18007](https://issues.apache.org/jira/browse/CASSANDRA-18007)
CircleCI improvements and parity - this is an epic which I hope it addresses your concerns, if it doesn't feel free to add whatever is missing to the epic. Thank you in advance

File: build.xml
Comment (ekaterinadimitrova2) [inline build.xml:186]: Unneeded new line, please, feel free to remove it on commit :-)

File: test/simulator/test/org/apache/cassandra/simulator/test/ShortPaxosSimulationTest.java
Comment (ekaterinadimitrova2) [inline test/simulator/test/org/apache/cassandra/simulator/test/ShortPaxosSimulationTest.java:43]: So this was the change to trigger the repeated tests, thanks. You can remove it on commit :-)

**Pull request closed without merging**
